cache/redis_cache.go

```diff
@@ -106,76 +98,222 @@ func (r *redisCache) Get(key *Key) (*CachedData, error) {
+		Ttl:             ttl,
+	}
 
-	if err != nil {
-		log.Errorf("Not able to fetch TTL for: %s ", key)
+	return value, nil
+}
+*/
+
+func (r *redisCache) Get(key *Key) (*CachedData, error) {
+	ctx, cancelFunc := context.WithTimeout(context.Background(), getTimeout)
+	defer cancelFunc()
+	nbBytesToFetch := int64(100 * 1024)
+	// fetching 100kBytes from redis to be sure to have the full metadata and,
+	//  for most of the queries that fetch a few data, the cached results
+	val, err := r.client.GetRange(ctx, key.String(), 0, nbBytesToFetch).Result()
+	// if key not found in cache
+	if errors.Is(err, redis.Nil) || len(val) == 0 {
+		return nil, ErrMissing
 	}
 
-	decoded, err := base64.StdEncoding.DecodeString(payload.Payload)
+	// others errors, such as timeouts
 	if err != nil {
-		log.Errorf("failed to decode payload: %s , due to: %v ", payload.Payload, err)
+		log.Errorf("failed to get key %s with error: %s", key.String(), err)
 		return nil, ErrMissing
 	}
+	stringKey := key.String()
+	ttl, err := r.client.TTL(ctx, stringKey).Result()
+	if err != nil {
+		return nil, fmt.Errorf("failed to ttl of key %s with error: %w", key.String(), err)
+	}
+	b := []byte(val)
+	metadata, offset := r.metadataFromByte(b)
+	if (int64(offset) + metadata.Length) < nbBytesToFetch {
+		// the condition is true ony if the bytes fetched contain the metadata + the cached results
+		// so we extract from the remaining bytes the cached results
+		payload := b[offset:]
+		reader := &io_reader_decorator{Reader: bytes.NewReader(payload)}
+		value := &CachedData{
+			ContentMetadata: *metadata,
+			Data:            reader,
+			Ttl:             ttl,
+		}
+
+		return value, nil
+	}
+	// since the cached results in redis are too big, we can't fetch all of them because of the memory overhead.
+	// we will create an io.reader that will fetch redis bulk by bulk to reduce the memory usage
+	reader := NewRedisStreamReader(uint64(offset), r.client, stringKey)
 	value := &CachedData{
-		ContentMetadata: ContentMetadata{
-			Length:   payload.Length,
-			Type:     payload.Type,
-			Encoding: payload.Encoding,
-		},
-		Data: bytes.NewReader(decoded),
-		Ttl:  ttl,
+		ContentMetadata: *metadata,
+		Data:            &io_reader_decorator{Reader: reader},
+		Ttl:             ttl,
 	}
 
 	return value, nil
 }
 
-func (r *redisCache) Put(reader io.Reader, contentMetadata ContentMetadata, key *Key) (time.Duration, error) {
-	data, err := toBytes(reader)
-	if err != nil {
-		return 0, err
-	}
+// this struct is here because CachedData requires an io.ReadCloser
+// but logic in the the Get function generates only an io.Reader
+type io_reader_decorator struct {
+	io.Reader
+}
 
-	encoded := base64.StdEncoding.EncodeToString(data)
-	payload := &redisCachePayload{
-		Length: contentMetadata.Length, Type: contentMetadata.Type, Encoding: contentMetadata.Encoding, Payload: encoded,
-	}
+func (m io_reader_decorator) Close() error {
+	return nil
+}
+func (r *redisCache) stringToBytes(s string) []byte {
```

Comment:
I think it's more than `stringToBytes`. The way I understand it, this function:
- on the first 4 bytes encodes len of the string
- appends the string to the slice
Can we call it `encodeString` and leave the comment? Reading this function take a moment

Reply:
Actually I just realised that it's the same code as in filesystem cache, isn't it? (both encode and decode - there it's called writeHeader and readHeader, maybe that's a better naming. Except that there we write to writer and read from reader)

Reply:
it is really transforming a string to a byte[] in order for serialization usage, I'll rename it to encodeString (but we will lose the information that the string becomes a byte array). Yes the logic is closed to writeHeader/readHeader execpt that it's more granulare since we're only dealing with a single string and not the headers.